### PR TITLE
EVG-7925 split up event-metajob

### PR DESCRIPTION
**Pull request by john-m-liu**
- Move dispatching events out of event-metajob into its new one. It still sends unsent notifications
- New job takes a batch (size controlled by existing throttle setting) of events and does processing for them
- Query for events is single-threaded so that it does not dispatch the same event more than once. It now finds all unprocessed events and dispatches them
- Dispatcher is multithreaded/distributed among all apps
- Any events missed (errors, shutdowns) will be picked up by the next run

### CODE REVIEW

File: units/event_metajob.go [hunk 1]
```diff
@@ -339,3 +190,8 @@ func (j *eventMetaJob) logEventCount() error {
 
 	return nil
 }
+
+func sha256sum(ids []string) string {
```
Comment (tychoish) [inline units/event_metajob.go:197]: sort the strings so you get a stable hash

File: units/event_notifier.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+	return j
+}
+
+func NewEventNotifierJob(env evergreen.Environment, q amboy.Queue, id string, events []string) amboy.Job {
+	j := makeEventNotifierJob()
+	j.q = q
+	j.env = env
+
+	j.SetID(fmt.Sprintf("%s:%s", eventNotifierName, id))
```
Comment (tychoish) [inline units/event_notifier.go:62]: s/:/./

File: units/event_notifier.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+		}()
+	}
+	wg.Wait()
+
+	endTime := time.Now()
+	totalDuration := endTime.Sub(startTime)
+
+	grip.Info(message.Fields{
+		"job_id":     j.ID(),
+		"job":        eventMetaJobName,
```
Comment (tychoish) [inline units/event_notifier.go:158]: j.Type().Name rather than the const, and maybe change job to job_type

File: units/event_metajob.go [hunk 4]
```diff
@@ -302,19 +146,26 @@ func (j *eventMetaJob) Run(ctx context.Context) {
...
 	if len(j.events) == 0 {
 		return
 	}
-
-	j.AddError(j.dispatchLoop(ctx))
+	eventIDs := []string{}
+	for i, evt := range j.events {
+		eventIDs = append(eventIDs, evt.ID)
+		if (i+1)%j.limit == 0 || i+1 == len(j.events) {
+			j.q.Put(ctx, NewEventNotifierJob(j.env, j.q, sha256sum(eventIDs), eventIDs))
```
Comment (tychoish) [inline units/event_metajob.go]: I might just pass eventIDs here, and put the hashing in the constructor for the notifier job.

File: units/event_metajob.go [hunk 5]
```diff
@@ -302,19 +146,26 @@ func (j *eventMetaJob) Run(ctx context.Context) {
 	if j.limit <= 0 {
 		j.limit = evergreen.DefaultEventProcessingLimit
 	}
-	j.events, err = event.FindUnprocessedEvents(j.limit)
+	j.events, err = event.FindUnprocessedEvents(-1)
 	if err != nil {
 		j.AddError(err)
 		return
 	}
+	grip.Info(j.events)
```
Comment (tychoish) [inline units/event_metajob.go]: I wouldn't log this without more context

File: units/event_metajob.go [hunk 6]
```diff
@@ -302,19 +146,26 @@ func (j *eventMetaJob) Run(ctx context.Context) {
 	if j.limit <= 0 {
 		j.limit = evergreen.DefaultEventProcessingLimit
 	}
-	j.events, err = event.FindUnprocessedEvents(j.limit)
+	j.events, err = event.FindUnprocessedEvents(-1)
```
Comment (tychoish) [inline units/event_metajob.go:149]: why not lift all of this into a Populate() function in crons, it doesn't seem like generating the jobs is really going to take that long and this just seems confusing.

Comment (john-m-liu) [reply units/event_metajob.go]: Done. The event-metajob is no more!

File: units/event_metajob.go (hunk 1 above)
Comment (john-m-liu) [reply units/event_metajob.go:294]: done

File: units/event_notifier.go (hunk 2 above)
Comment (john-m-liu) [reply units/event_notifier.go]: done

File: units/event_notifier.go (hunk 3 above)
Comment (john-m-liu) [reply units/event_notifier.go]: done

File: units/crons.go [hunk 7]
```diff
@@ -213,10 +215,63 @@ func PopulateEventAlertProcessing(env evergreen.Environment, parts int) amboy.Qu
...
 
-		return errors.Wrap(queue.Put(ctx, NewEventMetaJob(env, queue, ts)), "failed to queue event-metajob")
+		return nil
 	}
+
+}
+
+// dispatchUnprocessedNotifications gets unprocessed notifications
+// leftover by previous runs and dispatches them
+func dispatchUnprocessedNotifications(ctx context.Context, q amboy.Queue, flags *evergreen.ServiceFlags) error {
```
Comment (tychoish) [inline units/crons.go:264]: it seems like this function is only called once? I'm not sure that having a seperate function is needed, and I'm a little bit iffy about having functions other than the populate function in crons.go

File: units/crons.go [hunk 8]
```diff
@@ -213,10 +215,63 @@ func PopulateEventAlertProcessing(env evergreen.Environment, parts int) amboy.Qu
...
+	unprocessedNotifications, err := notification.FindUnprocessed()
+	if err != nil {
+		return errors.Wrap(err, "can't find unprocessed notifications")
+	}
+
+	return dispatchNotifications(ctx, unprocessedNotifications, q, flags)
+}
+
+func PopulateEventAlertProcessing(env evergreen.Environment) amboy.QueueOperation {
```
Comment (tychoish) [inline units/crons.go:273]: I'd just name ProcessEvents "PopulateEventAlertProcessing and avoid the extra step (mostly for debugability.)

File: units/crons.go [hunk 9]
```diff
@@ -213,10 +215,63 @@ func PopulateEventAlertProcessing(env evergreen.Environment, parts int) amboy.Qu
 			return nil
 		}
 
-		ts := utility.RoundPartOfHour(parts).Format(TSFormat)
+		err = dispatchUnprocessedNotifications(ctx, queue, flags)
+		grip.Error(message.WrapError(err, message.Fields{
```
Comment (tychoish) [inline units/crons.go:219]: should we return here?

File: units/crons.go (hunk 8 above)
Comment (john-m-liu) [reply units/crons.go:273]: oops done

**Pull request merged**